modepy/quadrature/finding.py

```diff
@@ -0,0 +1,422 @@
+    matrix = np.zeros((n, n), dtype=np.complex128)
+    for i, f in enumerate(lcfunctions):
+        ncommon = min(len(basis), len(f.functions))
+        assert basis[:ncommon]  == f.functions[:ncommon]
+        basis.extend(f.functions[ncommon:])
+
+        ncoeff = len(f.coefficients)
+        matrix[i, :ncoeff] = f.coefficients
+
+    return LinearCombinationIntegrand(coefficients @ matrix, basis)
+
+
+def _mass_matrix(
+            integrate: Callable[[Integrand], np.inexact],
+            basis: Sequence[Integrand],
+        ) -> np.ndarray:
+    n = len(basis)
+    integrals = [
+        [
+            integrate(_ProductIntegrand((basis[i], _ConjugateIntegrand(basis[j]))))
+            for j in range(i+1)
+            ]
+        for i in range(n)
+    ]
+
+    # Let np.array deal with figuring out the result dtype
+    return np.array([
+        # Fill out the full matrix from the triangle computed above
+        [integrals[i][j] if i >= j else integrals[j][i] for j in range(n)]
+        for i in range(n)
+    ])
+
+
+def orthogonalize_basis(
+            integrate: Callable[[Integrand], np.inexact],
+            basis: Sequence[Integrand],
+        ) -> Sequence[LinearCombinationIntegrand]:
+    r"""
+    Let :math:`\Omega\subset\mathbb C^n` be a domain. (Domains
+    over the reals are allowable as well.) Returns linear combinations
+    of functions in *basis* that is orthogonal under the (complex-valued)
+    :math:`L^2` inner product induced by *integrate*.
+
+    :arg integrate: Computes an integral of the passed integrand over
+        :math:`\Omega`. Must use integration nodes compatible with
+        *basis*.
+    :arg basis: A sequence of functions that accept an array of nodes
+        of shape either ``(ndim, nnodes)`` or ``(nnodes,)`` and return
+        an array of shape ``(nnodes,)`` with the value of the
+        basis function at the node.
+    """
+    n = len(basis)
+    mass_mat = _mass_matrix(integrate, basis)
+
+    l_factor = la.cholesky(mass_mat)
+
+    from scipy.linalg import solve_triangular
+    l_inv = solve_triangular(l_factor, np.eye(n), lower=True)
+
+    assert la.norm(np.triu(l_inv, 1), "fro") < 1e-14
+
+    return [
+        linearly_combine(l_inv[i, :i+1], basis[:i+1])
+        for i in range(n)
+    ]
+
+
+@dataclass(frozen=True)
+class _ComplexToNDAdapter:
+    function: Integrand
+
+    def __call__(self, points: np.ndarray) -> np.ndarray:
+        rpoints  = np.array([points.real, points.imag])
+        return self.function(rpoints)
+
+
+def adapt_2d_integrands_to_complex_arg(
+            functions: Sequence[Integrand]
+        ) -> Sequence[Integrand]:
+    r"""Converts a list of :data:`Integrand`\ s taking nodes in real-valued
+    arrays of shape ``(2, nnodes)`` to ones accepting a single
+    complex-valued array of shape ``(nnodes,)``. See :func:`guess_nodes_vr`
+    for the main intended use case.
+    """
+    return [_ComplexToNDAdapter(f) for f in functions]
+
+
+def guess_nodes_vr(
```

Comment:
```suggestion
def vioreanu_rokhlin_nodes_guess(
```
?

Reply:
I've ended up with `guess_nodes_vioreanu_rokhlin`. I find that "lead with a verb" is a good guideline for function names.

Reply:
I was thinking of guess as a noun :grin: Looks good!